### DISCUSSION TITLE
compiling error

### DISCUSSION THREAD

**Discussion by ggwarrier** [kachurovskiy/nanoels#122] [Q&A]
'SemaphoreHandle_t' does not name a type
I am new to this type of things.  My poor eyesight also gets in the way.  
So what does this mean?
mutex is seen many times but where is it defined?

**Comment by kachurovskiy** [comment]
Hi @ggwarrier, make sure to first select `ESP32S3 Dev Module` as your target platform at the top of the Arduino IDE.

![image](https://github.com/kachurovskiy/nanoels/assets/517919/fbabcecd-1480-407d-b798-050b5dfe0fb0)

![image](https://github.com/kachurovskiy/nanoels/assets/517919/2cad967b-1b1c-46fb-9b0a-5a2372eae9a8)

Hope it helps.

**Comment by kachurovskiy** [comment]
> So what does this mean?
> mutex is seen many times but where is it defined?

Those are simply mutexes. You can read more about FreeRTOS logic used here at https://www.freertos.org/xSemaphoreCreateBinary.html

**Comment by ggwarrier** [comment]
Thank you, Sir.  Meanwhile I came across another interesting article -
https://www.youtube.com/watch?v=WyG4LOG8EfA.  Have you seen this.  I feel
envy about Mr. TonyFoale- he looks to be about my age and seems to be able
to do so many!

regards,
Gopal

**Reply by kachurovskiy** [reply]
Yes, I think I've seen it before but took another look. That's a very nice build.

**Reply by ggwarrier** [reply]
What impressed me was his "reckless confidence of youth"!  And the thread cutting on a taper.

**Comment by ggwarrier** [comment]
I was able to compile the h4 without errors
Sketch uses 348461 bytes (26%) of program storage space. Maximum is 1310720
bytes.
Global variables use 21708 bytes (6%) of dynamic memory, leaving 305972
bytes for local variables. Maximum is 327680 bytes.

Thank you Maxim.  Fantastic when you even included the images.  Besides
using ESP32S3 Dev module as the target I had to install the core32 library
and PREFERENCES.h.

Now I am waiting for your 4-axis version.  Out of curiosity I wanted to
play around and learn a little on my own and so ordered couple of ESP32S
Dev mcu from Amazon.

Again thanks a bunch, Maxon.

regards,
Gopal

**Comment by ggwarrier** [comment]
Hello Maxim,
He uses a Mega for his UI and ESP32 for encoder counting.  FYI he lives in
Spain.
Coming to my situation, I compiles the code succesfully but I was not sure
the actual mcu that I should use.  I tried ESP32 WROOM-32D but came up with
the following message
Wrote 363728 bytes (208361 compressed) at 0x00010000 in 3.1 seconds
(effective 949.9 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
Is this wrong?  What is the actual MCU used in your design?

Meanwhile I am waiting on my order A1025.  Thank you
Gopal

**Reply by kachurovskiy** [reply]
Hi Gopal,

> He uses a Mega for his UI and ESP32 for encoder counting. 

Perhaps with older ESP32s it was harder to combine everything into one - or maybe there's some other reason.

```
Leaving...
Hard resetting via RTS pin...
```

This means that software is uploaded successfully and you just need to disconnect the USB and reconnect it back for the software to boot. 

Cheers!

**Comment by ggwarrier** [comment]
Hi Maxim,
It's past 10pm for you and you are so prompt in responding.  Thank you.

While I wait for the 4-axis box from you, I am going to make a prototype
board and try your design - just out of curiosity.  Wish me luck.

regards,
Gopal

**Reply by kachurovskiy** [reply]
Gopal, good luck! Do you already have the encoder and motors attached to the lathe?

**Comment by ggwarrier** [comment]
Maxim,  I have the encoder, ball screws  and the stepper motors and servo
motors in hand.  Since I only have basic tools to work with mounting these
on to my lathe is a challenge to me but I will get there.  I fell ill
recently (heart attack that required surgery) I am also physically
handicapped for the time being.

I noticed that you have modified your encoder mounting 3D parts.  I have an
old Ender 3 printer.  Will you be willing to share the stl file for that?

Again thank you.
regards,
Gopal

**Reply by kachurovskiy** [reply]
Gopal, that's a warrior spirit 👍 You can find the motor mount info below:

- https://github.com/kachurovskiy/nanoels/blob/main/hardware.md
- https://www.thingiverse.com/thing:4714722/remixes (original + remixes by other people, belt connection is really preferable there)

**Reply by ggwarrier** [reply]
Hello Maxim,

My Lathe has the spindle gear too close to the bulk head that I am unable to use the gear mounting you suggested.  I am trying to figure out a design on Tinker Cad to suit my lathe.  Being half blind I have gone thru a few with too many measurement errors.  I am using infill of 10% not to waste printer filamnet and time!
By the way I received your 4 axis control box and was able to connect the encoder and the stepper driver and the motor on my desk and was able to move the motor on both the axes.  I connected the encoder (600ppr) to a small dc motor to turn it.  
The big issue I now have is to amke any sense of all the modes you have on the box.  I tried to learn some of it by watching your videos but must admit that I am totally lost.
Do you have by any chance a user manual for the 4 axis controller.  Obviously I will only be using the Z and X axis for now.  My aim is to use it to do profiling similar to https://www.youtube.com/watch?v=hyncFl1DY4Q 
Thank you for all your help and putting up with these na·ive questions.
Gopal

**Reply by kachurovskiy** [reply]
Gopal, you can find the manual at https://github.com/kachurovskiy/nanoels/tree/main/h4

> My aim is to use it to do profiling similar to https://www.youtube.com/watch?v=hyncFl1DY4Q

This will be easily possible with https://kachurovskiy.github.io/lathecode/ and it already works in experimental mode - though there are some problems I need to work out before declaring it ready for use (support stop during operation, fix a few bugs, etc.).
